test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Actions/UIActionEngineTests.cs

```diff
@@ -181,6 +197,101 @@ public void AddUiActionEngineTelemetryProperties_AddsVulnerabilityInfo_Succeeds(
+            var dictMetadata = new Dictionary<string, object>
+            {
+                [NuGetProjectMetadataKeys.UniqueName] = "a",
+                [NuGetProjectMetadataKeys.ProjectId] = "a"
+            };
+            ProjectMetadataContextInfo metadataCtxtInfo = ProjectMetadataContextInfo.Create(dictMetadata);
+            prjMgrSvc
+                .Setup(mgr => mgr.GetMetadataAsync(It.IsAny<string>(), It.IsAny<CancellationToken>()))
+                .Returns(new ValueTask<IProjectMetadataContextInfo>(metadataCtxtInfo));
+            var uiService = new Mock<INuGetUI>();
+            var uiContext = new Mock<INuGetUIContext>();
+            var projectContext = new Mock<INuGetProjectContext>();
+            var serviceBroker = new Mock<IServiceBroker>();
+            _ = serviceBroker.Setup(sb => sb.GetProxyAsync<INuGetProjectManagerService>(
+                    It.Is<ServiceRpcDescriptor>(s => s == NuGetServices.ProjectManagerService),
+                    It.IsAny<ServiceActivationOptions>(),
+                    It.IsAny<CancellationToken>()))
+                .Returns(new ValueTask<INuGetProjectManagerService>(prjMgrSvc.Object));
+            uiContext.Setup(ctx => ctx.ServiceBroker).Returns(serviceBroker.Object);
+            uiService.Setup(ui => ui.UIContext).Returns(uiContext.Object);
+            uiService.Setup(ui => ui.ProjectContext).Returns(projectContext.Object);
+            uiService.Setup(ui => ui.Settings).Returns(settings.Object);
+            uiService.Setup(ui => ui.Projects).Returns(new[] { new ProjectContextInfo("a", ProjectModel.ProjectStyle.PackageReference, NuGetProjectKind.PackageReference) });
+
+            var action = UserAction.CreateInstallAction(packageIdToInstall, NuGetVersion.Parse("1.0.0"), isSolutionLevel, activeTab);
+
+            // Act
+            await uiEngine.PerformInstallOrUninstallAsync(uiService.Object, action, CancellationToken.None);
+
+            // Assert
+            Assert.NotNull(lastTelemetryEvent);
+            // expect failed action because we mocked just enough objects to emit telemetry
+            Assert.Equal(NuGetOperationStatus.Failed, lastTelemetryEvent["Status"]);
+            Assert.Equal(NuGetOperationType.Install, lastTelemetryEvent[nameof(ActionsTelemetryEvent.OperationType)]);
+            Assert.Equal(isSolutionLevel, lastTelemetryEvent[nameof(VSActionsTelemetryEvent.IsSolutionLevel)]);
+            Assert.Equal(activeTab, lastTelemetryEvent[nameof(VSActionsTelemetryEvent.Tab)]);
+
+            bool hasValue = false;
+            var enumerator = lastTelemetryEvent.GetEnumerator();
+            while (enumerator.MoveNext())
+            {
+                hasValue = enumerator.Current.Key == nameof(VSActionsTelemetryEvent.IsPackageToInstallTransitive);
+                if (hasValue)
+                {
+                    break;
+                }
+            }
+            Assert.Equal(containsValue, hasValue);
+            if (containsValue)
```

Comment:
As a general rule, avoid "if" conditions in tests. This creates a branch in the code that could be missing out on testing one path. In this case, for any of the test inputs that we supply that have containsValue, which is the expected value, set to false, we won't assert. We should instead assert that the expected value is what we expect it to be under all test circumstances. If we need to only assert under one condition and not under another, then what we need is a separate unit test to cover that other scenario.

Reply:
Removed if condition and enumeration iteration. Replaced with three state bool paramether `expectedValue`. Thanks!